src/client/websocket/WebSocketManager.js

```diff
@@ -233,17 +233,17 @@ class WebSocketManager extends EventEmitter {
         }
       });
 
-      shard.on(ShardEvents.INVALID_SESSION, () => {
+      const reconnectShard = (() => {
         this.client.emit(Events.SHARD_RECONNECTING, shard.id);
-      });
+        this.shardQueue.add(shard);
+        this.reconnect();
+      }).bind(this);
+
+      shard.on(ShardEvents.INVALID_SESSION, reconnectShard);
```

Comment:
Thanks a lot for finding an edge case with WS (again ;;)! However, this isn't the right fix imho. Honestly, if we get an invalid session that's nit resumable, we should just destroy the shard and clean it up, unless other's have any other suggestions 👀

Reply:
I mean, it is the Discord docs that say that upon receiving this opcode we should resume/reconnect: https://discord.com/developers/docs/topics/opcodes-and-status-codes

![image](https://user-images.githubusercontent.com/14080148/97313671-924eb400-186f-11eb-8799-91c6f58e9c06.png)

The way it works currently, is that the connection is lost, the shard is not properly destroyed and the connection is kept intact, while discord has destroyed the session on their end, leaving the shard in a sort of limbo where it keeps sending packets but receives nothing from Discord.

Reply:
Can we test this? I mean, the docs want us to close connection, then re-identify/resume... This would keep the conn alive (and just re-identify).. Is there even any way to test this?

Reply:
I did test this.

1. Apply only commit 43218347f0ddad180a9fe67e1b08c8065c0928d6
2. Do `someGuild.members.fetch({ user: ['a', '1'] })`
3. Watch the client hang for a few seconds then receive `Invalid Session`

With my changes it seems to reconnect alright after that.